### PR TITLE
Fix bugs in WeekOfMonth.apply, Week.onOffset

### LINKED ISSUE #18510: Week.onOffset looks fishy
A Week offset with weekday=None behaves very much like a Tick offset.  The main difference is how they handle `onOffset`.  Assuming `normalize` is False, the `Week` object will always return `week.onOffset(other)` --> `False` while the `Day` object will always return `day.onOffset(other)` --> `True`.

Is this intentional?

### LINKED ISSUE #18672: WeekOfMonth vs LastWeekOfMonth inconsistency
https://github.com/pandas-dev/pandas/blob/master/pandas/tseries/offsets.py#L1496

WeekOfMonth and LastWeekOfMonth have small differences in how they construct getOffsetOfMonth and then in what they do with it.  It looks fishy.  Any idea why these are done differently?  If there's not a reason for it, I'd like to align+merge them.

```
class WeekOfMonth(DateOffset):
    [...]
    @apply_wraps
    def apply(self, other):
        base = other
        offsetOfMonth = self.getOffsetOfMonth(other)

        months = self.n
        if months > 0 and offsetOfMonth > other:
            months -= 1
        elif months <= 0 and offsetOfMonth < other:
            months += 1

        other = self.getOffsetOfMonth(shift_month(other, months, 'start'))
        other = datetime(other.year, other.month, other.day, base.hour,
                         base.minute, base.second, base.microsecond)
        return other

    def getOffsetOfMonth(self, dt):
        w = Week(weekday=self.weekday)
        d = datetime(dt.year, dt.month, 1, tzinfo=dt.tzinfo)
        # TODO: Is this DST-safe?
        d = w.rollforward(d)
        return d + timedelta(weeks=self.week)


    def onOffset(self, dt):
        if self.normalize and not _is_normalized(dt):
            return False
        d = datetime(dt.year, dt.month, dt.day, tzinfo=dt.tzinfo)
        return d == self.getOffsetOfMonth(dt)

class LastWeekOfMonth(DateOffset):
    [...]
    @apply_wraps
    def apply(self, other):
        offsetOfMonth = self.getOffsetOfMonth(other)

        months = self.n
        if months > 0 and offsetOfMonth > other:
            months -= 1
        elif months <= 0 and offsetOfMonth < other:
            months += 1

        return self.getOffsetOfMonth(shift_month(other, months, 'start'))

    def getOffsetOfMonth(self, dt):
        m = MonthEnd()
        d = datetime(dt.year, dt.month, 1, dt.hour, dt.minute,
                     dt.second, dt.microsecond, tzinfo=dt.tzinfo)
        eom = m.rollforward(d)
        # TODO: Is this DST-safe?
        w = Week(weekday=self.weekday)
        return w.rollback(eom)

    def onOffset(self, dt):
        if self.normalize and not _is_normalized(dt):
            return False
        return dt == self.getOffsetOfMonth(dt)

```

### LINKED ISSUE #18864: BUG: WeekOfMonth.apply, LastWeekOfMonth.apply
Related: #18672 

I am now reasonably confident that `WeekOfMonth.apply` and `LastWeekOfMonth.apply` both have a bug caused by using `self.getOffsetOfMonth(dt)` instead of `self._get_offset_day(dt)

```
>>> offset = pd.offsets.WeekOfMonth(n=-2, week=2, weekday=3)
>>> ts = pd.Timestamp('2017-12-21 19:32:49.533991002')
>>> offset.onOffset(ts)
True
>>> (ts + offset) - offset ==  ts   # should be the same as offset.onOffset(ts)
False
>>> ts + offset
Timestamp('2017-11-16 19:32:49.533991002')
```

But given that `offset.onOffset(ts)` is `True`, we would expect `ts + offset` to move back by 2 months, not 1.

Surfacing this bug requires timestamps with non-zero nanos and n<0.  I expect there are other cases that would surface with tzaware timestamps.

Not that hard to fix; I'd like to get #18762 in first to avoid merge conflicts.

### PR DESCRIPTION
**Pull request by jbrockmendel**
In the process we get rid of `WeekOfMonth.getOffsetOfMonth` and `LastWeekOfMonth.getOffsetOfMonth`, which were idiosyncratic what arguments they passed to `datetime`.

The issues this addresses are orthogonal to #18762, but the code affected does overlap.  In particular, after this, `roll_monthday` will not be needed in 18762.

closes #18864
closes #18672 
closes #18510
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.23.0 [hunk 1]
```diff
@@ -362,3 +362,4 @@ Other
 
 - Improved error message when attempting to use a Python keyword as an identifier in a ``numexpr`` backed query (:issue:`18221`)
 - Bug in :class:`Timestamp` where comparison with an array of ``Timestamp`` objects would result in a ``RecursionError`` (:issue:`15183`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0]: pls move both this and the one higher to conversion. pls dont' put things in Other that fit elsewhere.

File: pandas/tseries/offsets.py [hunk 2]
```diff
@@ -112,6 +112,31 @@ def wrapper(self, other):
...
+    -------
+    shifted: datetime or Timestamp
+    """
+    if other.tzinfo is None:
+        return other + timedelta(days=days)
+
+    tz = other.tzinfo
+    naive = other.replace(tzinfo=None)
+    shifted = naive + timedelta(days=days)
+    return tslib._localize_pydatetime(shifted, tz)
```
Comment (jreback) [inline pandas/tseries/offsets.py]: side issue. ``_localize_pydatetime`` is in the wrong place (its in tslib) should be in conversion.pyx. And It shouldn't just return on ``None`` as well, and needs a better doc-string. can you add to list.

File: pandas/tseries/offsets.py [hunk 3]
```diff
@@ -1400,34 +1449,11 @@ def __init__(self, n=1, normalize=False, week=None, weekday=None):
...
-        d = w.rollforward(d)
-        return d + timedelta(weeks=self.week)
-
-    def onOffset(self, dt):
-        if self.normalize and not _is_normalized(dt):
-            return False
-        d = datetime(dt.year, dt.month, dt.day, tzinfo=dt.tzinfo)
-        return d == self.getOffsetOfMonth(dt)
+    def _get_offset_day(self, other):
+        mstart = datetime(other.year, other.month, 1)
```
Comment (jreback) [inline pandas/tseries/offsets.py]: can you document

File: pandas/tseries/offsets.py [hunk 4]
```diff
@@ -1482,31 +1508,12 @@ def __init__(self, n=1, normalize=False, weekday=None):
...
-        # TODO: Is this DST-safe?
-        w = Week(weekday=self.weekday)
-        return w.rollback(eom)
-
-    def onOffset(self, dt):
-        if self.normalize and not _is_normalized(dt):
-            return False
-        return dt == self.getOffsetOfMonth(dt)
+    def _get_offset_day(self, other):
+        dim = ccalendar.get_days_in_month(other.year, other.month)
```
Comment (jreback) [inline pandas/tseries/offsets.py]: same

File: doc/source/whatsnew/v0.23.0 [hunk 5]
```diff
@@ -362,3 +362,4 @@ Other
...
+- Bug in :class:`WeekOfMonth` and class:`Week` where addition and subtraction did not roll correctly (:issue:`18672`,:issue:`18510`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0]: there is a 3rd issue no?

**Review by jreback — CHANGES_REQUESTED** [review]
comments. looks good after changes. ping on green.

File: doc/source/whatsnew/v0.23.0 (hunk 1 above)
Comment (jbrockmendel) [reply doc/source/whatsnew/v0.23.0]: OK.  Out of curiosity, what is the heuristic that gets this lumped into "conversion"?

**Comment by jbrockmendel** [comment]
Ping

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**